### PR TITLE
Contact owner

### LINKED ISSUE #723: requesting ownership
It would be nice if a dandiset had an option to request ownership.

### PR DESCRIPTION
**Pull request by marySalvi**
Closes #723 

Allows a user to send an e-mail populated with a subject 'Regarding  Dandiset `name` `id`' to either the contact person or owner(if logged in)

### CODE REVIEW

File: web/src/views/DandisetLandingView/DownloadDialog.vue
Comment (marySalvi) [inline web/src/views/DandisetLandingView/DownloadDialog.vue:5]: This comment is in regards to the the following `v-menu`. Just wanted to call attention to it on whether it should be changed or not. No need for it to be done in this PR or at all.

File: web/src/views/DandisetLandingView/ContactDialog.vue
Comment (waxlamp) [inline web/src/views/DandisetLandingView/ContactDialog.vue:53]:
```suggestion
              Dandiset Owners
```

File: web/src/views/DandisetLandingView/ContactDialog.vue
Comment (waxlamp) [inline web/src/views/DandisetLandingView/ContactDialog.vue:44]: This should be refactored from here and the other place it appears into a computed property of the component. That will furthermore enable you to deal with possible missing data (all the optional chaining operators that appear) in a more robust way (either raising an error or supplying values besides `undefined` to use), as well as improve the readability of this expression.

Comment (waxlamp) [reply web/src/views/DandisetLandingView/ContactDialog.vue:44]: While you're at it, I think it's odd to have the body filled it with my own email address. I assume that can be left out and you'll just get your email client with a fresh, blank email body?

File: web/src/views/DandisetLandingView/DownloadDialog.vue
Comment (waxlamp) [reply web/src/views/DandisetLandingView/DownloadDialog.vue:5]: Why would we want to change this? If there's a problem with the width for this other menu, I would file an issue and/or open a different PR to discuss/deal with it.

Comment (waxlamp) [inline web/src/views/DandisetLandingView/DownloadDialog.vue:5]:
```suggestion
```

**Review by waxlamp — CHANGES_REQUESTED** [review]
Looking good! A few changes requested.

Also, I noticed that the title part of the dialog seems to be perhaps overly restricted by its width? Not sure how to describe it better than that but here's a picture:
overflow

This is happening at a pretty wide browser width, so I wonder if maybe the menu component needs a minimum width or something like that.

File: web/src/views/DandisetLandingView/DandisetActions.vue
Comment (waxlamp) [inline web/src/views/DandisetLandingView/DandisetActions.vue:48]: What happened here? Was this wrong before?

Comment (marySalvi) [reply web/src/views/DandisetLandingView/DandisetActions.vue:48]: Yes. I included it in this PR because I am adding to the Dandiset actions.

File: web/src/views/DandisetLandingView/DownloadDialog.vue
Comment (marySalvi) [reply web/src/views/DandisetLandingView/DownloadDialog.vue:5]: Definitely will be removing the comment, just looking to bring attention to this.

It is odd to have a set width instead of something responsive. It is even odder to have a set width in this manner: 
```
min-width:500px;
max-width:500px;
```

I am bringing attention to this now because the DownloadDialog is the component we chose to influence the design of the ContactDialog

Comment (marySalvi) [reply web/src/views/DandisetLandingView/DownloadDialog.vue:5]: explained below.

File: web/src/views/DandisetLandingView/ContactDialog.vue
Comment (marySalvi) [reply web/src/views/DandisetLandingView/ContactDialog.vue]: changed in 5f5b0cff9887ebefcd1e5640f1faccf5b9552afc

**Comment by marySalvi** [comment]
> Also, I noticed that the title part of the dialog seems to be perhaps overly restricted by its width? Not sure how to describe it 
> This is happening at a pretty wide browser width, so I wonder if maybe the menu component needs a minimum width or something like that.

fixed in 5f5b0cff9887ebefcd1e5640f1faccf5b9552afc

File: web/src/views/DandisetLandingView/ContactDialog.vue
Comment (marySalvi) [reply web/src/views/DandisetLandingView/ContactDialog.vue]: changed in d9847970259f7fafb446622621c4ae9ee8b30e6d

File: web/src/views/DandisetLandingView/DownloadDialog.vue
Comment (marySalvi) [reply web/src/views/DandisetLandingView/DownloadDialog.vue]: comment removed in 55db5cd4b43e05d463bc64f895b523248b54b653

File: dandiapi/api/views/dandiset.py
Comment (marySalvi) [inline dandiapi/api/views/dandiset.py:430]: adding to the users function throws a linting error that it is now too complex. I could use some guidance on how best to resolve this. @waxlamp

Comment (waxlamp) [reply dandiapi/api/views/dandiset.py:430]: One way to solve this is to always include `email` as a key in the response object, but set it to `None` for a unauthenticated user, and to the email address otherwise. That would do the following:
1. Eliminate the extra control flow complexity that leads to a higher [McCabe score](https://docs.astral.sh/ruff/rules/complex-structure/).
2. Make the response more uniform across the user's login state.

To do this you'd eliminate an if block in favor of a ternary conditional (something like `owner_user.email if request.user.is_authenticated else None`), but I don't know if that counts as a "decision point" to the McCabe algorithm.

Another take: the high complexity score is a clue that we were already doing a lot in this function, and have now tipped over into doing too much. Perhaps a cleaner solution is to add another endpoint that just gets the owner emails. This endpoint could just 403 for unauthenticated callers and unconditionally return email addresses otherwise. The cost would be another function in the codebase and another request at runtime. I'm not sure which approach is better.

A lot of uncertainty in my answer here. We can meet to hash this out -- let me know if you'd like to do so.

File: web/src/views/DandisetLandingView/ContactDialog.vue
Comment (waxlamp) [inline web/src/views/DandisetLandingView/ContactDialog.vue:103]: What is the behavior of the email client when hitting this code path?

File: web/src/views/DandisetLandingView/ContactDialog.vue
Comment (waxlamp) [inline web/src/views/DandisetLandingView/ContactDialog.vue:105]: I'm concerned about how this will break if `currentDandiset` is `null` or `undefined`. But I think in context of this component, that should never happen, which in turn makes me think we should be able to narrow the type of `currentDandiset` to eliminate `null | undefined` by making assertions and throwing an exception if the narrowing fails.

File: web/src/views/DandisetLandingView/ContactDialog.vue
Comment (waxlamp) [inline web/src/views/DandisetLandingView/ContactDialog.vue:108]:
```suggestion
```

File: web/src/views/DandisetLandingView/ContactDialog.vue
Comment (waxlamp) [inline web/src/views/DandisetLandingView/ContactDialog.vue:110]:
```suggestion
</script>

<style scoped>
```

File: web/src/views/DandisetLandingView/ContactDialog.vue
Comment (waxlamp) [inline web/src/views/DandisetLandingView/ContactDialog.vue:96]: The `.toString()` here might be better expressed as an explicit conversion to the string needed for the mailto link. `Array.toString()` happens to do the right thing, but the coincidence indicates that explicit would be better than implicit (to borrow from the Zen of Python).

Also, you might consider leaving this as an array of email addresses, to be converted to a string closer to the point of use (i.e., in the `makeTemplate()` function).

**Review by waxlamp — CHANGES_REQUESTED** [review]

File: dandiapi/api/views/dandiset.py
Comment (marySalvi) [reply dandiapi/api/views/dandiset.py:36]: Yes this was ruff

File: dandiapi/api/views/dandiset.py
Comment (marySalvi) [reply dandiapi/api/views/dandiset.py:411]: yes

File: web/src/views/DandisetLandingView/ContactDialog.vue
Comment (marySalvi) [reply web/src/views/DandisetLandingView/ContactDialog.vue:44]: using disabled also disables hover which would disable the tooltip. Setting selectable keeps the hover intact without allowing it to be clicked. We could do something else here to be more direct.

File: web/src/views/DandisetLandingView/ContactDialog.vue
Comment (marySalvi) [reply web/src/views/DandisetLandingView/ContactDialog.vue:103]: Throw error here

File: web/src/views/DandisetLandingView/ContactDialog.vue
Comment (marySalvi) [reply web/src/views/DandisetLandingView/ContactDialog.vue]: addressed in c3b06bbcc2361e74fd359ec2c0f80cfba0b02e99

**Comment by mvandenburgh** [comment]
One important thing to point out about this is that we're now going to be exposing user emails in the public API. With this change, anyone can retrieve a DANDI user's email through this endpoint. Previously, user emails were only available to admins (unless of course they manually added it to the dandiset metadata).

I'm not suggesting this is something that should block this PR, but I think it's important to note.

File: web/src/views/DandisetLandingView/ContactDialog.vue
Comment (waxlamp) [inline web/src/views/DandisetLandingView/ContactDialog.vue:119]: In the deploy preview for this PR, this is working for the "contact person" option, but not for "dandiset owners". Example: in [Dandiset 200560](https://deploy-preview-1840--gui-staging-dandiarchive-org.netlify.app/dandiset/200560), the mailto link includes a blank email address (it reads `mailto:?subject=[...]`).

File: web/src/views/DandisetLandingView/ContactDialog.vue
Comment (waxlamp) [inline web/src/views/DandisetLandingView/ContactDialog.vue:117]: The subject line ends up rendering as `Regarding Dandiset NeuroConv SDK Testing DANDI:200560/draft` which feels jumbled up. A better template might be:

```suggestion
    const subject = encodeURIComponent(`Regarding Dandiset ${currentDandiset.value.dandiset.identifier} ("${currentDandiset.value.name}")`);
```

This names the Dandiset by its identifier (i.e. its six-digit, zero-padded ID) which is normally how we refer to Dandisets, and puts its name in quotes within parentheses. So more like `Regarding Dandiset 200560 ("NeuroConv SDK Testing")`.

**Review by waxlamp — CHANGES_REQUESTED** [review]
Found some bugs, and I think we can simplify some of the `undefined | null` processing to make the template code clearer.

**Comment by marySalvi** [comment]
> One important thing to point out about this is that we're now going to be exposing user emails in the public API. With this change, anyone can retrieve a DANDI user's email through this endpoint. Previously, user emails were only available to admins (unless of course they manually added it to the dandiset metadata).
> 
> I'm not suggesting this is something that should block this PR, but I think it's important to note.

Agreed that this is an important issue to note. This concern was brought up to @satra  and the decision was that if a person is not logged in it is fine for them to access the contact person's email but they would need to be logged in for the owner's email.

**Comment by marySalvi** [comment]
> I saw that you accepted the spacing suggestion between the <script> and <style> sections. Do you want to apply this one too?

Actually I just resolved it after our conversation.

File: web/src/views/DandisetLandingView/ContactDialog.vue
Comment (marySalvi) [inline web/src/views/DandisetLandingView/ContactDialog.vue:130]: @waxlamp I pulled out most of the disabled logic and I _think_ this addresses most of your concerns. Let me know if this is not what you want

File: web/src/views/DandisetLandingView/ContactDialog.vue
Comment (waxlamp) [reply web/src/views/DandisetLandingView/ContactDialog.vue:120]: This is due to the deploy preview going to the production API, which does not have the changes from this PR that are needed to correctly populate the owners list.

File: web/src/views/DandisetLandingView/ContactDialog.vue
Comment (waxlamp) [inline web/src/views/DandisetLandingView/ContactDialog.vue:129]:
```suggestion
const isDisabled = (contacts: string[] | undefined): boolean =>
  contacts === undefined || contacts.length === 0 || contacts[0] === '';
```

**Review by mvandenburgh — APPROVED** [review]
I added a couple minor changes around simplifying types and moving some boolean logic out of the template and into the script section.

**Comment by mvandenburgh** [comment]
I assumed the "auto-merge" button would have waited for CI to pass :frowning_face: ... this is only going to staging so worse case I can just revert if it doesn't pass

**Pull request merged**
